AK/SipHash.cpp

```diff
@@ -0,0 +1,168 @@
+ */
+
+#include <AK/MemoryStream.h>
+#include <AK/SipHash.h>
+#include <AK/Span.h>
+#include <AK/UFixedBigInt.h>
+
+#ifdef KERNEL
+#    include <Kernel/Security/Random.h>
+#else
+#    include <AK/Random.h>
+#endif
+
+namespace AK {
+
+#define ROTL(x, b) (u64)(((x) << (b)) | ((x) >> (64 - (b))))
+
+#define SIPROUND           \
+    do {                   \
+        v0 += v1;          \
+        v1 = ROTL(v1, 13); \
+        v1 ^= v0;          \
+        v0 = ROTL(v0, 32); \
+        v2 += v3;          \
+        v3 = ROTL(v3, 16); \
+        v3 ^= v2;          \
+        v0 += v3;          \
+        v3 = ROTL(v3, 21); \
+        v3 ^= v0;          \
+        v2 += v1;          \
+        v1 = ROTL(v1, 17); \
+        v1 ^= v2;          \
+        v2 = ROTL(v2, 32); \
+    } while (0)
+
+template<size_t message_block_rounds, size_t finalization_rounds>
+static void do_siphash(ReadonlyBytes input, u128 key, Bytes output)
+{
+    VERIFY((output.size() == 8) || (output.size() == 16));
+
+    u64 v0 = 0x736f6d6570736575ull;
+    u64 v1 = 0x646f72616e646f6dull;
+    u64 v2 = 0x6c7967656e657261ull;
+    u64 v3 = 0x7465646279746573ull;
+    auto const left = input.size() & 7;
+    // The end of 64-bit blocks.
+    auto const block_end = input.size() - (input.size() % sizeof(u64));
+    u64 b = input.size() << 56;
+    v3 ^= key.high();
+    v2 ^= key.low();
+    v1 ^= key.high();
+    v0 ^= key.low();
+
+    if (output.size() == 16)
+        v1 ^= 0xee;
+
+    FixedMemoryStream input_stream { input };
+    for (size_t input_index = 0; input_index < block_end; input_index += 8) {
+        // The way we use them, no stream functions can fail.
+        auto const m = MUST(input_stream.read_value<LittleEndian<u64>>());
+        v3 ^= m;
+
+        for (size_t i = 0; i < message_block_rounds; ++i)
+            SIPROUND;
+
+        v0 ^= m;
+    }
+
+    switch (left) {
+    case 7:
+        b |= (static_cast<u64>(input[block_end + 6])) << 48;
+        [[fallthrough]];
+    case 6:
+        b |= (static_cast<u64>(input[block_end + 5])) << 40;
+        [[fallthrough]];
+    case 5:
+        b |= (static_cast<u64>(input[block_end + 4])) << 32;
+        [[fallthrough]];
+    case 4:
+        b |= (static_cast<u64>(input[block_end + 3])) << 24;
+        [[fallthrough]];
+    case 3:
+        b |= (static_cast<u64>(input[block_end + 2])) << 16;
+        [[fallthrough]];
+    case 2:
+        b |= (static_cast<u64>(input[block_end + 1])) << 8;
+        [[fallthrough]];
+    case 1:
+        b |= (static_cast<u64>(input[block_end + 0]));
+        break;
+    case 0:
+        break;
+    }
+
+    v3 ^= b;
+
+    for (size_t i = 0; i < message_block_rounds; ++i)
+        SIPROUND;
+
+    v0 ^= b;
+
+    if (output.size() == 16)
+        v2 ^= 0xee;
+    else
+        v2 ^= 0xff;
+
+    for (size_t i = 0; i < finalization_rounds; ++i)
+        SIPROUND;
+
+    b = v0 ^ v1 ^ v2 ^ v3;
+
+    FixedMemoryStream output_stream { output };
```

Comment:
And this is probably...about 10% of your perf loss.

Reply:
Same as above; see the reference for how ugly this gets otherwise.

Reply:
I don't see it, I wasn't even trying and it looks just fine without:
```diff
@@ -113,8 +113,8 @@ static void do_siphash(ReadonlyBytes input, u128 key, Bytes output)
 
     b = v0 ^ v1 ^ v2 ^ v3;
 
-    FixedMemoryStream output_stream { output };
-    MUST(output_stream.write_value<LittleEndian<u64>>(b));
+    LittleEndian<u64> b_le { b };
+    output.overwrite(0, &b_le, sizeof(b_le));
 
     if (output.size() == 8)
         return;
@@ -125,7 +125,8 @@ static void do_siphash(ReadonlyBytes input, u128 key, Bytes output)
         SIPROUND;
 
     b = v0 ^ v1 ^ v2 ^ v3;
-    MUST(output_stream.write_value<LittleEndian<u64>>(b));
+    b_le = b;
+    output.overwrite(sizeof(b_le), &b_le, sizeof(b_le));
 }
 
 // Using a shared key greatly simplifies DynamicLoader and Kernel,
```